### ISSUE TITLE
[HOLD for payment 2024-01-11] [$1000] Draft message doesn't remain when It is empty

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#25488]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to any chat
2. Send new message
3. Edit this message
4. Replace the message by empty string (don't click save)
5. Click another report
6. Back to previous report
## Expected Result: 
Draft message should be remained
Note that: When replacing draft message by another string, It will be remained
## Actual Result: 
Draft message doesn't remain

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.55-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/21cba4ea-a809-42bd-af7f-384443df330f

https://github.com/Expensify/App/assets/43996225/95d38886-ce30-4881-b7cb-8e3143f1d906

**Expensify/Expensify Issue URL:**
**Issue reported by:** @dukenv0307 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1691551143871559

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01adda4a6239a4a1af</li>
        <li>Upwork Job ID: 1693680071403327488</li>
        <li>Last Price Increase: 2023-09-11</li>
    </ul>
</details>

**Comment by flaviadefaria** [comment]
@greg-schroeder  I'm logging off now and will be OoO for the next week so reassigning this to you in the meantime.  Thanks!

**Comment by greg-schroeder** [comment]
Ok, sure!

**Comment by greg-schroeder** [comment]
Hmmmmm. I don't know about this. Why would a user replace a message with an empty string? That doesn't seem like a viable use case to me. I don't know that we should optimize for that.

**Comment by greg-schroeder** [comment]
Started a discussion

**Comment by greg-schroeder** [comment]
Team agreed it's worth moving over to External

**Comment by mollfpr** [comment]
@samh-nl I followed through with your proposal, but it could have resulted better.

![Screenshot 2023-08-22 at 19 19 14](https://github.com/Expensify/App/assets/25520267/565d3996-88be-471b-997f-5fee1155e130)

The current user with a bunch of `reportActionsDrafts_` saved in their storage will be affected, so I don't think changing it to `undefined` will be a good idea.

**Comment by mollfpr** [comment]
@samh-nl Could we use the current key `reportActionsDrafts_`? If possible, we should avoid adding a new key and use the associate key.

**Comment by mollfpr** [comment]
Not overdue, it still open for proposals.

**Comment by greg-schroeder** [comment]
Awaiting new proposals

**Comment by greg-schroeder** [comment]
Still awaiting proposals, might send this to CS/SWM soon

**Comment by s-alves10** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Draft message doesn't remain when it's empty

### What is the root cause of that problem?

We save draft while it's changing here
https://github.com/Expensify/App/blob/b57d3a3d347aff057edb0bba2a4cd65448f4d670/src/pages/home/report/ReportActionItemMessageEdit.js#L198-L203

As you can see, when draft is empty, we store the original value. So when we remount the `ReportActionItemMessageEdit` component, original message would appear

This is the root cause

### What changes do you think we should make in order to solve the problem?

We can store the `space` character if draft is empty and trim it when mounting.
In the above line 202
```
        debouncedSaveDraft(' ');
```

And change the line
https://github.com/Expensify/App/blob/b57d3a3d347aff057edb0bba2a4cd65448f4d670/src/pages/home/report/ReportActionItemMessageEdit.js#L104
with
```
        return Str.htmlDecode(props.draftMessage.trim());
```

I think this change wouldn't have any side effects

<details>
    <summary>Result</summary>

https://github.com/Expensify/App/assets/126839717/a94b68d9-29b4-4398-9a2e-b1c9a9c19df0

</details>


### What alternative solutions did you explore? (Optional)

**Comment by greg-schroeder** [comment]
Bump @mollfpr for proposal review

**Comment by ShogunFire** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

When we edit a comment, delete all chars change report and comeback the draft input is filled with the former text instead of empty string

### What is the root cause of that problem?

We intentionally set the old string to the text input if it's empty because the condition for the Edit input to show up is that the draft message is not empty. We do that here: https://github.com/Expensify/App/blob/b57d3a3d347aff057edb0bba2a4cd65448f4d670/src/pages/home/report/ReportActionItemMessageEdit.js#L198-L203


### What changes do you think we should make in order to solve the problem?

We should not replace empty string by the old text.

In the collection ONYXKEYS.COLLECTION.REPORT_ACTIONS_DRAFTS we currently only store strings, we should also store a boolean isEditing. We must change this line https://github.com/Expensify/App/blob/d2bdaece9151c8c65d3b4bc4c9578be663f1950d/src/ONYXKEYS.ts#L384

with this:
`[ONYXKEYS.COLLECTION.REPORT_ACTIONS_DRAFTS]: Record<string, boolean>;`

Of course right now the messages stored in onyx don't have this boolean, we will set it to false but for compatibility we will have to test both conditions like this:

```
if (props.isEditing || !props.draftMessage)
    show Edit Input
```


### What alternative solutions did you explore? (Optional)

**Comment by mollfpr** [comment]
Will review the proposals in the morning.

**Comment by ShogunFire** [comment]
Just a note: I agree with this: https://github.com/Expensify/App/issues/25488#issuecomment-1699632670 , my proposal is if a migration is difficult/impossible

**Comment by mollfpr** [comment]
@samh-nl I'm unsure if it is worth doing, and I've never seen us doing it. We can make it work by creating a new key and eventually removing the old one.

@s-alves10 That might be work, but it's clearly a workaround.

>Just a note: I agree with this: https://github.com/Expensify/App/issues/25488#issuecomment-1699632670 , my proposal is if a migration is difficult/impossible

@ShogunFire Could you elaborate on this? Are your solution require a migration too?

**Comment by s-alves10** [comment]
@mollfpr 

We're now treating empty string as an action, not as a draft
We're treating spaces as an empty in draft already.

So I suggested to use space as an empty draft. This doesn't require onyx migration and requires minimal changes. Also this doesn't define a new state of messages.
I don't think this is a workaround

Note: We need to distinguish empty draft and non-draft. There are two ways to do this: one suggested by @samh-nl - needs onyx migration, and one described in my proposal. I think we should not introduce new state as possible.

**Comment by ShogunFire** [comment]
@mollfpr my proposal doesn't require a migration because we would still be testing the current condition: if the draft string is not empty then we show the edit mode. We would just add another condition with a new boolean isEditing, part of  the existing collection. Inspired by what you suggested here: https://github.com/Expensify/App/issues/25488#issuecomment-1688174202

**Comment by greg-schroeder** [comment]
Thoughts on the above @mollfpr?

**Comment by mollfpr** [comment]
>I don't think this is a workaround

@s-alves10 How it's not a workaround? A whitespace is not an empty string; an empty string should have no character.

---

@ShogunFire How the `isEditing` will be set? Will changing the value types cause the error log when the draft value is only a string?

---

@samh-nl I believe you already have the answer to the Onyx migration.

>The alternative I outlined would be to add a separate boolean value to store which are being edited, for example reportActionsEditing_. For backwards compatibility, we could set editing = true if draft is not empty (or isEditing = editing || !empty(draft)). I can work out the implementation details more if you think it may be a good approach.

Could this solution be implemented? I'm okay with adding a new key now since we don't have any clean approach.

**Comment by mollfpr** [comment]
>The alternative I outlined would be to add a separate boolean value to store which are being edited, for example reportActionsEditing_. For backwards compatibility, we could set editing = true if draft is not empty (or isEditing = editing || !empty(draft)). I can work out the implementation details more if you think it may be a good approach.

@samh-nl  I'm inclined to see this as a proposal, if you will.

**Comment by mollfpr** [comment]
> Please give the other option of renaming the Onyx key also some consideration, it's a bit more elegant and the trade-off (existing drafts are removed when people refresh) may not be a problem in practice.

If you can explain the clear plan, you can start a discussion in the slack regarding the trade of.

**Comment by greg-schroeder** [comment]
Looks like conversation continues on trying to find the right proposal

**Comment by greg-schroeder** [comment]
Bumped slack discussion

**Comment by mollfpr** [comment]
Not overdue. I will revisit the proposals and the slack discussion.

**Comment by greg-schroeder** [comment]
Slack discussion hasn't really gone anywhere. What do you want to do here @mollfpr?

**Comment by mollfpr** [comment]
Sorry for the delay 🙏

> Please give the other option of renaming the Onyx key also some consideration, it's a bit more elegant and the trade-off (existing drafts are removed when people refresh) may not be a problem in practice.

@greg-schroeder I like the above suggestion better, but the trade-off might need to be conveyed to the internal.

**Comment by greg-schroeder** [comment]
Okay. I bumped the internal team

**Comment by greg-schroeder** [comment]
@samh-nl can you summarize where we ended up here?

**Comment by greg-schroeder** [comment]
Excellent recap, thank you. Let's hold then.

**Comment by greg-schroeder** [comment]
No update yet

**Comment by greg-schroeder** [comment]
holding

**Comment by greg-schroeder** [comment]
@samh-nl Looks like the issue we were holding for was finished up - do you think we're good to proceed here?

**Comment by mollfpr** [comment]
Sorry for the delay.

The proposal looks good, but I still don't know when and how the migration will be run. @samh-nl Could you elaborate on the migration step?

**Comment by greg-schroeder** [comment]
Bump @samh-nl!

**Comment by greg-schroeder** [comment]
Does that help answer your question @mollfpr?

**Comment by mollfpr** [comment]
>Migrations are defined in [this](https://github.com/Expensify/App/tree/main/src/libs/migrations) directory and subsequently called [here](https://github.com/Expensify/App/blob/main/src/libs/migrateOnyx.js)

Got it! The [proposal](https://github.com/Expensify/App/issues/25488#issuecomment-1684067167) looks good to me!

🎀 👀 🎀 C+ reviewed!

**Comment by greg-schroeder** [comment]
This is on staging, awaiting deploy to prod

**Comment by greg-schroeder** [comment]
You can ignore this @sonialiap, it's a bug, I will handle payment

**Comment by mollfpr** [comment]
>[@mollfpr] The PR that introduced the bug has been identified. Link to the PR:
 [@mollfpr] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:

No offending PR.

> [@mollfpr] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:

The regression test should be good.

> [@mollfpr] Determine if we should create a regression test for this bug.
 [@mollfpr] If we decide to create a [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) for the bug, please propose the regression test steps to ensure the same bug will not reach production again.

1. Open any report
2. Send a message
3. Press the Edit button in the newly added message
4. Empty the draft message input
5. Navigate to a different report
6. Return to the first report
7. **_Verify_** that the draft message input is empty
8. 👍 or 👎 

---

@greg-schroeder No need to send me the offer; I'll make a manual request in NewDot. Thank you!

**Comment by greg-schroeder** [comment]
reporter: $50
c: $500
c+: $500


Offers sent to @dukenv0307 and @samh-nl, and you can do a manual request @mollfpr.

**Comment by greg-schroeder** [comment]
filing regression test

**Comment by JmillsExpensify** [comment]
$500 payment approved for @mollfpr (c+) based on summary above.

**Comment by greg-schroeder** [comment]
Sorry all, let me review this one

**Comment by greg-schroeder** [comment]
Hmm. This is tricky. The issue was reported/created in August before our first pricing changes, but no PR was created until late 2023, well after our payscale changed. I think we could probably honor the original reporting bonus given the date, but the work done on the PR took place well after the $500 baseline was established. That seems most fair to me.
